pkg/resources/digitalocean/dns/dns.go

```diff
@@ -0,0 +1,501 @@
+func (r *resourceRecordSets) Zone() dnsprovider.Zone {
+	return r.zone
+}
+
+// recordRecordSet implements dnsprovider.ResourceRecordSet which represents
+// a single record associated with a zone
+type resourceRecordSet struct {
+	name       string
+	data       string
+	ttl        int
+	recordType rrstype.RrsType
+}
+
+// Name returns the name of a resource record set
+func (r *resourceRecordSet) Name() string {
+	return r.name
+}
+
+// Rrdatas returns a list of data associated with a resource record set
+// in DO this is almost always the IP of a record
+func (r *resourceRecordSet) Rrdatas() []string {
+	return []string{r.data}
+}
+
+// Ttl returns the time-to-live of a record
+func (r *resourceRecordSet) Ttl() int64 {
+	return int64(r.ttl)
+}
+
+// Type returns the type of record a resource record set is
+func (r *resourceRecordSet) Type() rrstype.RrsType {
+	return r.recordType
+}
+
+// resourceRecordChangeset implements dnsprovider.ResourceRecordChangeset
+type resourceRecordChangeset struct {
+	client *godo.Client
+	zone   *zone
+	rrsets dnsprovider.ResourceRecordSets
+
+	additions []dnsprovider.ResourceRecordSet
+	removals  []dnsprovider.ResourceRecordSet
+	upserts   []dnsprovider.ResourceRecordSet
+}
+
+// Add adds a new resource record set to the list of additions to apply
+func (r *resourceRecordChangeset) Add(rrset dnsprovider.ResourceRecordSet) dnsprovider.ResourceRecordChangeset {
+	r.additions = append(r.additions, rrset)
+	return r
+}
+
+// Remove adds a new resource record set to the list of removals to apply
+func (r *resourceRecordChangeset) Remove(rrset dnsprovider.ResourceRecordSet) dnsprovider.ResourceRecordChangeset {
+	r.removals = append(r.removals, rrset)
+	return r
+}
+
+// Upsert adds a new resource record set to the list of upesrts to apply
+func (r *resourceRecordChangeset) Upsert(rrset dnsprovider.ResourceRecordSet) dnsprovider.ResourceRecordChangeset {
+	r.upserts = append(r.upserts, rrset)
+	return r
+}
+
+// Apply adds new records stored in r.additions, updates records stored
+// in r.upserts and deletes records stored in r.removals
+func (r *resourceRecordChangeset) Apply() error {
+	if len(r.additions) > 0 {
+		for _, record := range r.additions {
+			recordCreateRequest := &godo.DomainRecordEditRequest{
+				Name: record.Name(),
+				Data: record.Rrdatas()[0],
+				TTL:  int(record.Ttl()),
+				Type: string(record.Type()),
+			}
+			err := createRecord(r.client, r.zone.Name(), recordCreateRequest)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if len(r.removals) > 0 {
+		records, err := getRecords(r.client, r.zone.Name())
+		if err != nil {
+			return err
+		}
+
+		for _, record := range r.removals {
+			var desiredRecord godo.DomainRecord
+			found := false
+			for _, domainRecord := range records {
+				if domainRecord.Name == record.Name() {
+					desiredRecord = domainRecord
+					found = true
+				}
+			}
+			if !found {
+				return errors.New("could not find desired record to remove")
```

Comment:
`fmt.Errorf` instead please.  We do not use `errors` anywhere in our code base.  Odd but heh.  @justinsb has pinged me on this before.